Reformat and add pre-commit checks

arichardson opened the pull request:
Also moves generated files to `contrib/` to avoid formatting them. PR for CI.

arichardson committed:
Move sail-derived code to a separate folder

This will make it easier to disable clang-format for these files and makes
it obvious that these files should not be edited since they are either
vendored from sail or generated.

arichardson committed:
Sync sail-provided headers with latest sail

Also remove the local rts.c and place all the code that it contained in
sail_wrapper_common.c

arichardson committed:
Add a basic pre-commit config file

arichardson committed:
Run pre-commit checks as part of CI

arichardson committed:
Add clang-format to the pre-commit CI

jrtc27 commented on test/generated/elf.h:
```diff
@@ -1 +1,75 @@
-// This is included by sail generated code but not actually needed
```
These files aren't generated?

arichardson replied:
They are derived from sail.

arichardson replied:
I just had a stub one here - a later commit in this series copies the sail version over.

jrtc27 replied:
Yes, but they're not generated, they're hand-written, and you're just vendoring them, so they don't belong in a directory named generated

arichardson replied:
What would you suggest instead?

arichardson replied:
I agree that generated isn't the best name - I just used that since the regen_* scripts create them.

jrtc27 replied:
Have a test/{contrib,vendor,3rdparty,...}?

arichardson replied:
Done

jrtc27 commented on cheri_compressed_cap.h:
```diff
@@ -46,9 +46,9 @@
-#include "cheri_compressed_cap_64.h"
 #include "cheri_compressed_cap_128.h"
 #include "cheri_compressed_cap_128m.h"
+#include "cheri_compressed_cap_64.h"
```
Ew...

arichardson replied:
I could disable include sorting here?

jrtc27 replied:
I think that would be better, yeah

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -111,11 +111,11 @@ _CC_STATIC_ASSERT(_CC_N(MAX_RESERVED_OTYPE) <= _CC_N(MAX_REPRESENTABLE_OTYPE), "
-static inline uint32_t _cc_N(get_perms)(const struct _cc_N(cap)* cap);
-static inline uint8_t _cc_N(get_reserved)(const struct _cc_N(cap)* cap);
-static inline uint32_t _cc_N(get_uperms)(const struct _cc_N(cap)* cap);
+static inline uint8_t _cc_N(get_flags)(const struct _cc_N(cap) * cap);
```
clang-format seems confused by these?

arichardson replied:
Yeah I know - I figured I'd just deal with the broken formatting rather than adding `//clang-format off`. Can do that if you prefer?

jrtc27 replied:
(Perhaps we should hoist the typedef and #define?)

arichardson replied:
Yes that's what I've done now.

jrtc27 commented on decompress_cap_128.c:
```diff
@@ -48,7 +48,7 @@ static const char* otype_suffix(uint32_t otype) {
 #define OTYPE_CASE(Name, ...)                                                                                          \
     case CC128_##Name: return " (CC128_" #Name ")";
-    CC128_LS_SPECIAL_OTYPES(OTYPE_CASE, )
+        CC128_LS_SPECIAL_OTYPES(OTYPE_CASE, )
```
This is worse

jrtc27 commented on test/fuzz_decompress_common.cpp:
```diff
@@ -49,8 +49,7 @@ static const char* otype_suffix(uint32_t otype) {
     case _CC_N(Name): return " (" STRINGIFY(_CC_N(Name)) ")";
-        _CC_N(LS_SPECIAL_OTYPES)(OTYPE_CASE, )
-    default: return "";
+        _CC_N(LS_SPECIAL_OTYPES)(OTYPE_CASE, ) default : return "";
```
This is particularly bad

arichardson committed:
Use _cc_cap_t earlier

The advantage of this is that the macro does not confuse clang-format so
it correctly formats the * as a pointer rather than multiplication

arichardson committed:
Reduce code duplication in fuzzer

We can reuse test_utils.h instead of duplicating those functions.

arichardson committed:
Reformat all files

arichardson committed:
Disable clang-format for third-party files

arichardson committed:
Move setbounds_inputs_128 to contrib/

This file was generated from sail output, no need to clang-format it.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -111,11 +111,13 @@ _CC_STATIC_ASSERT(_CC_N(MAX_RESERVED_OTYPE) <= _CC_N(MAX_REPRESENTABLE_OTYPE), "
 
 // Forward-declare the accessors since we use them inside the struct body:
 struct _cc_N(cap);
```
This is now implicitly done by the typedef so you don't need this line

jrtc27 commented on decompress_cap_128m.c:
```diff
@@ -45,12 +45,14 @@
 #define OTYPE_CASE(Name, ...)                                                                                          \
     case CC128M_##Name: return " (CC128M_" #Name ")";
-    CC128M_LS_SPECIAL_OTYPES(OTYPE_CASE, )
+        CC128M_LS_SPECIAL_OTYPES(OTYPE_CASE, )
```
Deindent

arichardson replied:
Not sure how that ended up in the diff. Fixed.

arichardson committed:
Use the _cc_bounds_bits to work around clang-format bug

Also looks slightly cleaner.

jrtc27 commented on CMakeLists.txt:
```diff
@@ -109,7 +109,7 @@ set_target_properties(GMP::GMP PROPERTIES
 
 function(add_sail_wrapper _format _suffix _sanitizer_flags)
-    add_library(sail_wrapper_${_format}${_suffix} STATIC test/sail_wrapper_${_format}.c test/contrib/sail.c test/contrib/rts.c test/contrib/sail_failure.c)
+    add_library(sail_wrapper_${_format}${_suffix} STATIC test/sail_wrapper_${_format}.c test/contrib/sail.c test/contrib/sail_failure.c)
```
Can we delete the file if we don't need it any more?

arichardson replied:
Sigh, I must have brought it back because I changed the name of the folder while rebasing. Thanks for all the review - hopefully done once I push that change...

jrtc27 commented on test/test_common.cpp:
```diff
@@ -1,7 +1,7 @@
 #include "../cheri_compressed_cap.h"
-#include <unistd.h>
```
?

arichardson replied:
No longer needed since the catch include provides all the necessary definitions.

jrtc27 replied:
I guess it was a workaround for it not being self-contained before then?.. Doesn't belong in a reformat commit though

arichardson replied:
Yeah I've moved it into the correct commit now.
